lib/Notifier.js

```diff
@@ -0,0 +1,370 @@
+    } 
+  }
+
+  /**
+   * User API
+   */
+  /**
+   * Please call after creating the Notifier instance and setting the event handlers.
+   */
+  start()
+  {  
+    logger.debug('start()');
+
+    this.receiveRequest(this._initial_subscribe);
+  }
+
+  /**
+   * Switch pending dialog state to active.
+   */
+  setActiveState() 
+  {
+    logger.debug('setActiveState()');
+
+    if (this._state === C.STATE_PENDING) 
+    {
+      this._state = C.STATE_ACTIVE;
+    }
+  }
+ 
+  /**
+   *  Send the initial and subsequent notify request.
+   *  @param {string} body - notify request body.
+   */
+  notify(body = null) 
+  {
+    logger.debug('notify()');
+
+    // Prevent send notify after final notify.
+    if (!this._dialog)
+    {
+      logger.warn('final notify has sent');
+
+      return;
+    }
+
+    let subs_state = this._stateNumberToString(this._state);
+ 
+    if (this._state !== C.STATE_TERMINATED) 
+    {
+      let expires = Math.floor((this._expires_timestamp - new Date().getTime()) / 1000);
+
+      if (expires < 0)
+      {
+        expires = 0;
+      }
+
+      subs_state += `;expires=${expires}`;
+    } 
+    else 
+    {
+      if (this._terminated_reason) 
+      {
+        subs_state += `;reason=${this._terminated_reason}`;
+      }
+      if (this._terminated_retry_after !== null)
+      {
+        subs_state += `;retry-after=${this._terminated_retry_after}`;    
+      }
+    }
+
+    const headers = this._headers.slice();
+
+    headers.push(`Subscription-State: ${subs_state}`);
+
+    if (body) 
+    {
+      headers.push(`Content-Type: ${this._content_type}`);
+    }
+  
+    this._dialog.sendRequest(JsSIP_C.NOTIFY, { 
+      body, 
+      extraHeaders  : headers,
+      eventHandlers : {
+        onRequestTimeout : () =>
+        {
+          this._dialogTerminated(C.NOTIFY_RESPONSE_TIMEOUT);
+        },
+        onTransportError : () =>
+        {
+          this._dialogTerminated(C.NOTIFY_TRANSPORT_ERROR);
+        },
+        onErrorResponse : (response) =>
+        {
+          if (response.status_code === 401 || response.status_code === 407)
+          {
+            this._dialogTerminated(C.NOTIFY_FAILED_AUTHENTICATION);
+          }
+          else 
+          {
+            this._dialogTerminated(C.NOTIFY_NON_OK_RESPONSE);
+          }
+        }, 
+        onDialogError : () =>
+        {
+          this._dialogTerminated(C.NOTIFY_NON_OK_RESPONSE);
+        } 
+      }
+    });
+  }
+
+  /**
+   *  Terminate. (Send the final NOTIFY request).
+   * 
+   * @param {string} body - Notify message body.
+   * @param {string} reason - Set Subscription-State reason parameter.
+   * @param {number} retryAfter - Set Subscription-State retry-after parameter.
+   */
+  terminate(body = null, reason = null, retryAfter = null) 
+  {
+    logger.debug('terminate()');
+    
+    this._state = C.STATE_TERMINATED;
+    this._terminated_reason = reason;
+    this._terminated_retry_after = retryAfter;
+
+    this.notify(body);
```

Comment:
I don't like `this._terminated_reason` and `this._terminated_retry_after` being set here and making the public method `this.notify()` deal with them.

I would rather create a private method `_sendNotify(body, headers)`, that is used from `terminate()` and from the public `notify()`. Then `notify()` can assert that current state is not TERMINATED, otherwise throw.

Reply:
added private _notify() method that called from terminate() and notify().
notify() throw exception if called in TERMINATED state.